### PR TITLE
[s2s] test_distributed_eval

### LINKED ISSUE #8297: [s2s] 1 GPU test for run_distributed_eval
Add test coverage for run_distributed_eval.py that can run on 1 GPU.
The command:
```bash
python -m torch.distributed.launch --nproc_per_node=1 run_distributed_eval.py --model_name Helsinki-NLP/opus-mt-en-ro --save_dir opus_wmt_en_ro_gens --data_dir wmt_en_ro
```

works on 1 GPU.

After adding test coverage, we could try to improve API consistency between run_distributed_eval.py and run_eval.py .

### PR DESCRIPTION
**Pull request by stas00**
This PR:

* [x] adds a helper function `get_gpu_count()`, which returns the number of available gpus (regardless of whether torch or tf is used) (otherwise we have to do all the `if _torch_available: import torch` rigmarole. 
* [x] adds a basic test for `run_distributed_eval.py`

Things I wasn't sure about:
- I put the new test into `test_seq2seq_examples_multi_gpu.py`, but it doesn't require multigpu. should I create a new test file  `test_seq2seq_examples_gpu.py` or just leave it here for now and refactor once we have more tests like this?
- I made it `@slow`, but wasn't 100% sure - it takes some 25s - probably too slow.

Fixes: #8297

@sshleifer

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
LGTM for the testing utility!

**Comment by LysandreJik** [comment]
Fyi the example tests don't currently run on the multi-gpu setup, as pretty much none of the examples tests were made to run on a multi-gpu setup.

We can look into adding them to the CI.

File: examples/seq2seq/test_seq2seq_examples_multi_gpu.py [hunk 1]
```diff
@@ -80,3 +87,31 @@ def convert(k, v):
...
         self.assertEqual(len(metrics["val"]), desired_n_evals)
+
+    @slow
+    @require_torch_gpu
+    def test_distributed_eval(self):
+        output_dir = self.get_auto_remove_tmp_dir()
+        args = f"""
+            --model_name Helsinki-NLP/opus-mt-en-ro
+            --save_dir {output_dir}
+            --data_dir test_data/wmt_en_ro
```
Comment (sshleifer) [inline examples/seq2seq/test_seq2seq_examples_multi_gpu.py:98]: `--task translation` for bleu scores.
`--n_obs 100 --num_beams=2` for (a) faster test, (b) coverage of config pass-through.




(also in your training test PR can use `--eval_beams=2`)

**Review by sshleifer — COMMENTED** [review]
Your async utils are powerful! 
How long does this take? I can upload a smaller model if that's helpful. (Won't affect coverage.)

**Review by sshleifer — APPROVED** [review]

**Comment by stas00** [comment]
> Fyi the example tests don't currently run on the multi-gpu setup, as pretty much none of the examples tests were made to run on a multi-gpu setup.
> 
> We can look into adding them to the CI.

Absolutely!

Do you prefer to start with a CI job that explicitly lists tests that were converted to run on multi-gpu (3 at the moment) or just run them all? It's all that `require_torch_mutigpu` plus ` that `require_torch_gpu` as it's flexible.

File: examples/seq2seq/test_seq2seq_examples_multi_gpu.py (hunk 1 above)
Comment (stas00) [reply examples/seq2seq/test_seq2seq_examples_multi_gpu.py:98]: > How long does this take? 

As mentioned in the OP - 25sec.

> I can upload a smaller model if that's helpful. (Won't affect coverage.)

We surely can experiment with that.

Comment (stas00) [reply examples/seq2seq/test_seq2seq_examples_multi_gpu.py:98]: suggestions integrated - thank you! went down from 24secs to 20sec, yay!

File: examples/seq2seq/test_seq2seq_examples_multi_gpu.py [hunk 2]
```diff
@@ -96,6 +96,9 @@ def test_distributed_eval(self):
             --model_name Helsinki-NLP/opus-mt-en-ro
             --save_dir {output_dir}
             --data_dir test_data/wmt_en_ro
+            --n_obs 100
```
Comment (sshleifer) [inline examples/seq2seq/test_seq2seq_examples_multi_gpu.py:99]:
```suggestion
```
this won't matter on test data, my bad

**Review by sshleifer — APPROVED** [review]

**Comment by sshleifer** [comment]
> Do you prefer to start with a CI job that explicitly lists tests that were converted to run on multi-gpu (3 at the moment) or just run them all? It's all that require_torch_mutigpu plus that `require_torch_gpu` as it's flexible.

I would send a PR with whatever you think makes most sense. Or wait until the correct answer is clearer. At this point you know pretty much as much about CI tradeoffs as I do :)

**Comment by stas00** [comment]
I will prepare a new PR for github CIs.

otherwise I think this PR is complete.

**Pull request merged**